Scrapyard server

surprisetalk opened the pull request:
Implemented a new `server_command` function to serve `.scrap` files over HTTP, allowing retrieval by path or hash. Added error handling for missing files. Included unit tests for server functionality in `scrapscript_tests.py` to ensure proper operation and error responses.

surprisetalk committed:
Add server command functionality and corresponding tests

Implemented a new `server_command` function to serve `.scrap` files over HTTP, allowing retrieval by path or hash. Added error handling for missing files. Included unit tests for server functionality in `scrapscript_tests.py` to ensure proper operation and error responses.

tekknolagi commented on scrapscript.py:
```diff
@@ -2482,6 +2482,64 @@ def flat_command(args: argparse.Namespace) -> None:
+        print(f"Error: {dir} is not a valid directory")
+        sys.exit(1)
+
+    scraps = {}
```
This looks like it can be split out into a separate function that loads a database/yard from disk. Then we can also use it to resolve local scraps from eval (later)

tekknolagi commented on scrapscript.py:
```diff
@@ -2482,6 +2482,64 @@ def flat_command(args: argparse.Namespace) -> None:
+    import http.server
+    import socketserver
+    import hashlib
+    import os.path
```
```suggestion
```

tekknolagi commented on scrapscript.py:
```diff
@@ -2482,6 +2482,64 @@ def flat_command(args: argparse.Namespace) -> None:
+
+    class ScrapHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
+        def do_GET(self) -> None:
+            path = self.path.lstrip("/")
+            if path in scraps:
+                self.send_response(200)
+                self.send_header("Content-Type", "application/scrap; charset=binary")
+                self.send_header("Content-Disposition", f'attachment; filename="{path}.scrap"')
+                self.send_header("Content-Length", str(len(scraps[path])))
+                self.end_headers()
+                self.wfile.write(scraps[path])
```
```suggestion
            path = self.path.lstrip("/")
            scrap = scraps.get(path)
            if scrap is not None:
                self.send_response(200)
                self.send_header("Content-Type", "application/scrap; charset=binary")
                self.send_header("Content-Disposition", f'attachment; filename="{path}.scrap"')
                self.send_header("Content-Length", str(len(scrap)))
                self.end_headers()
                self.wfile.write(scrap)
```

tekknolagi replied:
Oh also consider using `json.dumps` for proper quoting in Content-Disposition (:/)

tekknolagi commented on scrapscript.py:
```diff
@@ -2521,6 +2579,16 @@ def main() -> None:
+
+    yard_server = yard_subparsers.add_parser("server")
+    yard_server.set_defaults(func=server_command)
+    yard_server.add_argument("directory", type=str, help="Directory to serve")
```
`default="."`?

tekknolagi commented on scrapscript_tests.py:
```diff
@@ -4051,5 +4051,47 @@ def test_pretty_print_variant(self) -> None:
+        self.server_thread.daemon = True
+        self.server_thread.start()
+
+        time.sleep(0.1)
```
This is really sketchy. I think you can/should instead wait on it being ready some other way if possible

tekknolagi commented on scrapscript.py:
```diff
@@ -2482,6 +2482,64 @@ def flat_command(args: argparse.Namespace) -> None:
+                        scraps[rel_path_without_ext] = serialized
+                        logger.debug(f"Loaded {rel_path_without_ext}")
+                        file_hash = hashlib.md5(serialized).hexdigest()
+                        scraps[f"${file_hash}"] = serialized
```
I don't think we should put them in the same namespace unless you enforce that filenames cannot start with `$`

tekknolagi reviewed (approved):
Nice first commit. It would be good to see some design sketches of where this is going so that I can help you integrate it into the codebase

For example: will the server eventually take uploads?

In general, I would like to see some communication before dropping PRs over the wall, especially if they bring bigger changes. Creating an issue is a good way to do distributed development so that others (Chris, Abel) can chime in.

surprisetalk committed:
Update scrapscript.py

Co-authored-by: Max Bernstein <tekknolagi@gmail.com>

surprisetalk committed:
Update scrapscript.py

Co-authored-by: Max Bernstein <tekknolagi@gmail.com>

surprisetalk commented:
> Nice first commit. It would be good to see some design sketches of where this is going so that I can help you integrate it into the codebase
> 
> For example: will the server eventually take uploads?
> 
> In general, I would like to see some communication before dropping PRs over the wall, especially if they bring bigger changes. Creating an issue is a good way to do distributed development so that others (Chris, Abel) can chime in.

Thanks for the feedback! I'll follow up with a few issues where we can discuss implementation before I dive into future PRs

surprisetalk committed:
Refactor server command tests and improve scrap file handling

Updated `scrapscript_tests.py` to dynamically determine the host and port for server tests, enhancing reliability. Modified `scrapscript.py` to validate scrap file names using regex and adjusted content disposition header formatting. Defaulted the directory argument in the server command to improve usability.

surprisetalk committed:
Refactor imports and improve code formatting in tests

Updated `scrapscript_tests.py` to streamline import statements and enhance code readability. Added `urllib.request` import at the top and organized other imports in the `setUp` method for better clarity.

tekknolagi commented on scrapscript_tests.py:
```diff
@@ -4052,44 +4052,46 @@ def test_pretty_print_variant(self) -> None:
+            try:
+                with socket.create_connection((self.host, self.port), timeout=0.1) as s: 
+                    break
+            except Exception:
```
Can you make this Exception more specific? Otherwise a typo could cause AttributeError which would cause this to loop indefinitely instead of raising

tekknolagi commented on scrapscript_tests.py:
```diff
@@ -4051,5 +4052,51 @@ def test_pretty_print_variant(self) -> None:
+            s.bind(("127.0.0.1", 0))
+            self.host, self.port = s.getsockname()
+
+        args = argparse.Namespace(
```
Using argparse in a test indicates to me that there is an impedance mismatch in the API design. It indicates that there should be a differently-testable interface that doesn't rely on CLI arguments. Probably fine for now but in the future we'll likely want a `serve_yard` function with only positional parameters or something

tekknolagi commented on scrapscript.py:
```diff
@@ -2521,6 +2580,16 @@ def main() -> None:
+    yard_server.add_argument("directory", type=str, default=".", help="Directory to serve")
+    yard_server.add_argument("--host", type=str, default="127.0.0.1", help="Host to bind to")
+    yard_server.add_argument("--port", type=int, default=8080, help="Port to listen on")
+    yard_server.add_argument("--debug", action="store_true")
```
If you don't use `--debug` (`args.debug`), drop it

surprisetalk committed:
Refactor error handling in server connection tests and remove unused debug argument

Updated `scrapscript_tests.py` to specifically catch `ConnectionRefusedError` and `socket.timeout` exceptions for improved error handling during server connection tests. Removed the unused `--debug` argument from the server command in `scrapscript.py` to streamline argument parsing.

tekknolagi commented on scrapscript_tests.py:
```diff
@@ -4051,5 +4052,51 @@ def test_pretty_print_variant(self) -> None:
 
 
+class ServerCommandTests(unittest.TestCase):
+    def setUp(self) -> None:
```
Sorry, another thing I noticed: some tests are hanging because you never shut down the server. You should have whatever the equivalent of setUp is kill the thing

tekknolagi replied:
This is the only significant blocker; I can't merge until tests pass

tekknolagi commented on scrapscript.py:
```diff
@@ -2482,6 +2482,65 @@ def flat_command(args: argparse.Namespace) -> None:
+    scraps = {}
+    for root, _, files in os.walk(dir):
+        for file in files:
+            if re.match(r"^[a-zA-Z0-9-]+\.scrap$", file):
```
Nit: no need for `re`. If the filename starts with `$`, continue

tekknolagi commented on scrapscript.py:
```diff
@@ -2482,6 +2482,65 @@ def flat_command(args: argparse.Namespace) -> None:
+                        serialized = bytes(serializer.output)
+                        scraps[rel_path_without_ext] = serialized
+                        logger.debug(f"Loaded {rel_path_without_ext}")
+                        file_hash = hashlib.md5(serialized).hexdigest()
```
Why md5? The world moved on long ago

surprisetalk replied:
No particular reason. Would you prefer to start with `sha256`?

tekknolagi commented:
If you rebase on latest trunk, the unit test errors that are "my" fault go away

tekknolagi committed:
Don't use re

tekknolagi committed:
Explicitly shut down scrapyard server in tests

tekknolagi committed:
Print help for bare yard argument

tekknolagi committed:
Make yard server directory optional

tekknolagi committed:
Don't chdir

tekknolagi committed:
Don't bother passing absolute path

server_command uses abspath internally.

surprisetalk committed:
md5->sha256

surprisetalk commented:
Thanks for grabbing those fixes! LGTM 👍🏼 I'm unsure if you want to fix the failing deployments first, so feel free to merge at your leisure

tekknolagi commented:
Failing deployments are because you're the author, not me. It's a non-issue
